[IH-3] Add section links

chshersh opened the pull request:
Sorry for massive changes in imports and sorry for alignment :(

Imports are now sorted. No other changes were performed to them (because Serokell `stylish-haskell` is more close to current style of `ihdf`, only sorts imports, I don't think it's too bad so I need to disable `stylish-haskell` each time).

Regarding alignment: it might be a problem if a lot of people use different style and disabling alignment may decrease possible number of merge conflicts. But, as well, I don't think it hurts much for now...

As for feature: implementation works, I tested it on `traversable.html` chapter. Though code might not be really good, I'm seeing this project first time! Please, share your comments. I don't know how to check if section is present that's why I didn't do it :disappointed:

chshersh committed:
[IH-3] Add section links

int-index reviewed (changes requested):
It's important to have warnings for invalid references. Here's how I'd do it: instead of `MonadState [Warning]`, use `MonadState BookState`, where

```
data BookState = BookState
  { bookStateWarnings :: [Warning]
  , bookStateSectionsDefined :: Set SectionId
  , bookStateSectionsReferenced :: Set Section Id
  }
```

After parsing the chapter is done, compute a set difference and print a warning when it's non-empty.

chshersh committed:
[IH-3] Replace [Warning] with BookState

chshersh committed:
[IH-3] Use BookState to warn about invalid sections

chshersh commented:
I've implemented through `BookState`. But solution it's not working :( Because it sees `Shapes and Elements` section before this section becomes known.

int-index commented on src/BookParser.hs:
```diff
@@ -244,14 +267,22 @@ preprocessChapter = \case
 preprocessSection = \case
-  Span t -> return $ SectionRef $ SectionId t
+  Span t -> do
+     knownSections <- gets bookStateSectionsDefined
```
Why are you doing this in `preprocessSection`, not after the chapter was parsed entirely? This way forward referencing won't work.

chshersh committed:
[IH-3] Check sections after chapter parsed entirely

chshersh commented:
Tested. This now works as expected on `traversable.html` chapter.

int-index commented on src/BookParser.hs:
```diff
@@ -436,7 +431,16 @@ pSection = do
+  knownSections <- gets bookStateSectionsDefined
+  usedSections  <- gets bookStateSectionsReferenced
+
+  forM_ (Set.toList usedSections) $ \sectionId ->
```
Use set difference, `forM_ (Set.difference usedSections knownSections) $ \sectionId ->`. And you don't need `toList` because `Set` is `Foldable` (`forM_` uses any `Foldable`). And finally, I prefer `for_` to `forM_` because its type is less restrictive.

chshersh replied:
What about `for_` being inefficient for some data types?

int-index replied:
This was a bug because these types did not define a faster version of `(*>)` but did define a faster version of `(>>)`. I don't think it applies to `transformers` or `megaparsec` anymore.

int-index commented on src/BookParser.hs:
```diff
@@ -270,15 +270,10 @@ preprocessChapter = \case
+    let sId = SectionId t
+    sectionsReferenced %= Set.insert sId
+    return $ SectionRef sId
+  s      -> do
```
```
s      -> do
  ^^^^^
```
is this your idea of beautiful alignment? (I don't really mind, you can leave this line as is)

chshersh replied:
Yes, it was mine idea. At first it was just

```haskell
preprocessSection = \case
   Span t -> return $ SectionRef $ SectionId t
   s      -> do
```

But I forgot to remove alignment after this refactoring. Maybe I shouldn't introduce too much new alignment...

int-index reviewed (changes requested):
I asked specifically to use set difference when we discussed it, not to reimplement it inefficiently.

chshersh committed:
[IH-3] Use Set.difference

int-index reviewed (changes requested):
OK, final problem: when the section header is not `Span` but has some formatting, it can't be referenced.

I also don't like that `SectionId` contains full section name (like "A Section About Stuff") rather than its id (like "a-section-about-stuff"). Why did you decide to do this transformation in HTML rendering, not during parsing?

I suggest to add a function:

```
extractSectionId :: MonadBlah m => Span -> m (Span, SectionId)
```

Right now we can take the easy route and just expect a `Span` there, but very importantly -- throw a parse error if there's something different inside. Later we can improve it to strip formatting. And the `SectionId` it returns must contain an id like `"a-section-about-stuff"` -- preprocessed. This id can be stored in the parsed book and then used in HTML

chshersh committed:
[IH-3] Handle section references properly

Now `SectionId` stores id (i.e. "header-name" instead of "Header Name").
Also, utility functions which throws parser error when parsing section
is introduced.

int-index commented on src/BookStructure.hs:
```diff
@@ -14,13 +17,25 @@ incDepth (Depth n) = Depth (n+1)
+mkSectionId :: Text -> SectionId
+mkSectionId = SectionId . Text.intercalate "-" . Text.words . Text.toLower
+
+sectionIdToHeader :: SectionId -> Text
```
Are you kidding me?

int-index reviewed (changes requested):
I will finish this one myself

int-index committed:
Fix bad solution
